runtime/dynstats.c

```diff
@@ -499,12 +601,151 @@ dynstats_createCtr(dynstats_bucket_t *b, const uchar* metric, dynstats_ctr_t **c
+	if (b->stateFileDirectory == NULL) {
+		wrkdir = glblGetWorkDirRaw();
+	} else {
+		wrkdir = b->stateFileDirectory;
+	}
+	return(wrkdir);
+}
+
+/*
+*	Helper function to combine statefile and workdir
+*/
+static int getFullStateFileName(dynstats_bucket_t *b, uchar* pszstatefile, uchar* pszout, int buflen)
+{
+	int lenout;
+	const uchar* pszworkdir;
+
+	/* Get Raw Workdir, if it is NULL we need to propper handle it */
+	pszworkdir = getStateFileDir(b);
+
+	/* Construct file name */
+	lenout = snprintf((char*)pszout, buflen, "%s/%s",
+					(char*) (pszworkdir == NULL ? "." : (char*) pszworkdir), (char*)pszstatefile);
+
+	/* return out length */
+	return lenout;
+}
+
+/* this generates a state file name suitable for the given file. To avoid
+ * malloc calls, it must be passed a buffer which should be MAXFNAME large.
+ * Note: the buffer is not necessarily populated ... always ONLY use the
+ * RETURN VALUE!
+ * This function is guranteed to work only on config data and DOES NOT
+ * open or otherwise modify disk file state.
+ */
+static uchar * ATTR_NONNULL(1, 2)
+getStateFileName(const dynstats_bucket_t *const pbucket,
+		uchar *const __restrict__ buf,
+		const size_t lenbuf) {
+
+	DBGPRINTF("getStateFileName for '%s'\n", pbucket->name);
+	snprintf((char*)buf, lenbuf - 1, "dynstats-state:%s", pbucket->name);
+	DBGPRINTF("getStateFileName:  state file name now is %s\n", buf);
+	return buf;
+}
+
+/* This function persists dynstats_bucket_t data, which for the time being:
+ * metric bucket counters
+ */
+static rsRetVal ATTR_NONNULL(1)
+persistBucketState(dynstats_bucket_t *b) {
+	DEFiRet;
+	uchar statefile[MAXFNAME];
+	uchar statefname[MAXFNAME];
+	struct json_object *jval = NULL;
+	struct json_object *json = NULL;
+	struct json_object *json_bucket_values = NULL;
+
+	assert (b->name);
+	uchar *const statefn = getStateFileName(b, statefile, sizeof(statefile));
+	/* Get full path and file name */
+	getFullStateFileName(b, statefn, statefname, sizeof(statefname));
+	DBGPRINTF("persisting state for '%s', state file '%s'\n", b->name, statefname);
+
+	CHKmalloc(json = json_object_new_object());
+	jval = json_object_new_string((char*) b->name);
+	json_object_object_add(json, DYNSTATS_BUCKET_PERSIST_NAME, jval);
+
+	// walk the hashtable and create a val for each
+	CHKmalloc(json_bucket_values = json_object_new_object());
+	json_object_object_add(json, DYNSTATS_BUCKET_PERSIST_VALUES_NAME, json_bucket_values);
+
+	pthread_rwlock_rdlock(&b->lock);
+	if (hashtable_count(b->table) > 0) {
+		struct hashtable_itr *itr = hashtable_iterator(b->table);
+		dynstats_ctr_t *pctr = NULL;
+		do {
+			pctr = (dynstats_ctr_t*) hashtable_iterator_value(itr);
+			jval = json_object_new_int64(pctr->ctr);
+			json_object_object_add(json_bucket_values, (const char*)pctr->metric, jval);
+		} while (hashtable_iterator_advance(itr));
+		free(itr);
+	}
+	pthread_rwlock_unlock(&b->lock);
+
+	const char *jstr =  json_object_to_json_string_ext(json, JSON_C_TO_STRING_SPACED);
```

Comment:
For very large tables, `JSON_C_TO_STRING_PLAIN ` may be more efficient.

Reply:
yes.